moshi/src/main/java/com/squareup/moshi/JsonValueSource.kt

```diff
@@ -0,0 +1,208 @@
+ *
+ * It implements [lenient parsing][JsonReader.setLenient] and has no mechanism to
+ * enforce strict parsing. If the input is not valid or lenient JSON the behavior of this source is
+ * unspecified.
+ */
+internal class JsonValueSource @JvmOverloads constructor(
+  private val source: BufferedSource,
+  /** If non-empty, data from this should be returned before data from [source]. */
+  private val prefix: Buffer = Buffer(),
+  /**
+   * The state indicates what kind of data is readable at [limit]. This also serves
+   * double-duty as the type of bytes we're interested in while in this state.
+   */
+  private var state: ByteString = STATE_JSON,
+  /**
+   * The level of nesting of arrays and objects. When the end of string, array, or object is
+   * reached, this should be compared against 0. If it is zero, then we've read a complete value and
+   * this source is exhausted.
+   */
+  private var stackSize: Int = 0
+) : Source {
+  private val buffer: Buffer = source.buffer
+
+  /** The number of bytes immediately returnable to the caller. */
+  private var limit: Long = 0
+  private var closed = false
+
+  /**
+   * Advance [limit] until any of these conditions are met:
+   *  * Limit is at least `byteCount`. We can satisfy the caller's request!
+   *  * The JSON value is complete. This stream is exhausted.
+   *  * We have some data to return and returning more would require reloading the buffer. We prefer to return some
+   *  data immediately when more data requires blocking.
+   *
+   * @throws EOFException if the stream is exhausted before the JSON object completes.
+   */
+  private fun advanceLimit(byteCount: Long) {
+    while (limit < byteCount) {
+      // If we've finished the JSON object, we're done.
+      if (state === STATE_END_OF_JSON) {
+        return
+      }
+
+      // If we can't return any bytes without more data in the buffer, grow the buffer.
+      if (limit == buffer.size) {
+        if (limit > 0L) return
+        source.require(1L)
+      }
+
+      // Find the next interesting character for the current state. If the buffer doesn't have one,
+      // then we can read the entire buffer.
+      val index = buffer.indexOfElement(state, limit)
+      if (index == -1L) {
+        limit = buffer.size
+        continue
+      }
+      val b = buffer[index]
+      when {
+        state === STATE_JSON -> when (b.toInt().toChar()) {
+          '[', '{' -> {
+            stackSize++
+            limit = index + 1
+          }
+          ']', '}' -> {
+            stackSize--
+            if (stackSize == 0) state = STATE_END_OF_JSON
+            limit = index + 1
+          }
+          '\"' -> {
+            state = STATE_DOUBLE_QUOTED
+            limit = index + 1
+          }
+          '\'' -> {
+            state = STATE_SINGLE_QUOTED
+            limit = index + 1
+          }
+          '/' -> {
+            source.require(index + 2)
+            when (buffer[index + 1]) {
+              '/'.code.toByte() -> {
```

Comment:
does it make a difference, either from perf or readability, to coerce `buffer[index+1]` to `Char` instead of coercing `'/'` to a `Byte`?

Reply:
I don't think it helps with readability too much and we'd like to keep things at a byte